### ISSUE TITLE
Gradient calculate failed for flow past cylinder

### ISSUE THREAD

**Issue by zhangzhen117** [tum-pbs/PhiFlow#198]
Hi, I want to calculate the gradient wrt the forcing term in a flow past a cylinder problem. I tried the following simple code but get an error of TorchCustomFunctionBackward. The following are the details.

`
import phi
from phi.torch.flow import *
device = torch.device('cuda' if torch.cuda.is_available() else 'cpu')
print(device)
N          = 32                  # grid resolution 
x_min      = -5.0                # min X coordinate
x_max      = 15.0                # max X coordinate
y_min      = -5.0                # min Y coordinate
y_max      = 5.0                 # max Y coordinate
dt         = 0.1                # time step
nu         = 0.1                # kinematic viscosity
domain = Box(x=(x_min,x_max), y=(y_min,y_max))
cylinder = Box(x=(-0.5,0.5), y=(-0.5,0.5))
_=plot(domain, cylinder, overlay='args')
@jit_compile
def forward_run(forcing, v0, dt=dt, vis=nu):
    v = v0
    for i in range(2):
        v = advect.semi_lagrangian(v, v, dt)
        v = diffuse.explicit(v, vis, dt)
        v = v + forcing * dt
        v,_ = fluid.make_incompressible(v,cylinder, Solve('scipy-direct'))
    loss = field.l2_loss(v-field.stop_gradient(v0))
    return loss
boundary = {'x-': vec(x=1.0,y=0.0), 'x+': ZERO_GRADIENT, 'y': PERIODIC}
v0 = StaggeredGrid(vec(x=1.0,y=0.0), boundary=boundary, bounds=domain,x=N*2,y=N)
v0, _ = fluid.make_incompressible(v0, cylinder, Solve('scipy-direct')) 
_=plot(v0)
boundary = {'x-': 0, 'x+': ZERO_GRADIENT, 'y': PERIODIC}
forcing = StaggeredGrid(0, boundary=boundary, bounds=domain,x=N*2,y=N)
forward_run(forcing,v0)
sim_grad = field.gradient(forward_run,wrt='forcing',get_output=False)
forcing_grad = sim_grad(forcing,v0)
_=plot(forcing_grad)
`
